### PR TITLE
Register endpoint 2nd pass

### PR DESCRIPTION
**Pull request by susiecoleman**
- Add more specific error messages
 - Handle skip confirmation
 - Handle 3rd party T&Cs
 - Input validation

### CODE REVIEW

File: app/com/gu/identity/frontend/configuration/Configuration.scala [hunk 1]
```diff
@@ -23,5 +25,7 @@ class ApplicationConfiguration(val appConfiguration: PlayConfiguration) extends
   val identityApiKey = getString("identity.api.key")
   val identityApiHost = getString("identity.api.host")
 
+  val identityProfileBase = getString("identity.frontend.tandCsBase")
```
Comment (jamespamplin) [inline app/com/gu/identity/frontend/configuration/Configuration.scala]: Could this config key be named a bit simpler? ie: `identity.frontend.baseUrl`

File: conf/CODE.conf [hunk 2]
```diff
@@ -4,6 +4,8 @@ identity.api.host=idapi.code.dev-theguardian.com
 
 identity.frontend.cookieDomain=code.dev-theguardian.com
 
+identity.frontend.base="https://profile.code.dev-theguardian.com"
```
Comment (jamespamplin) [inline conf/CODE.conf]: Doesn't appear to be used. Probably should be `identity.frontend.tandCsBase` or `identity.frontend.baseUrl` as I've suggested above.

File: app/com/gu/identity/frontend/controllers/RegisterAction.scala [hunk 3]
```diff
@@ -41,21 +55,20 @@ class RegisterAction(identityService: IdentityService, val messagesApi: Messages
     NoCache {
       val clientIp = ClientRegistrationIp(request)
       registerForm.bindFromRequest.fold(
-        errorForm => Future.successful(SeeOther(routes.Application.register(Seq("error-registration"), None).url)),
+        errorForm => Future.successful(SeeOther(routes.Application.register(Seq("error-registration")).url)),
```
Comment (jamespamplin) [inline app/com/gu/identity/frontend/controllers/RegisterAction.scala]: Is there a way to surface which field has failed validation here? Will be useful to report this to the user.

File: app/com/gu/identity/frontend/configuration/Configuration.scala (hunk 1 above)
Comment (susiecoleman) [inline app/com/gu/identity/frontend/configuration/Configuration.scala]: Changed

File: conf/CODE.conf (hunk 2 above)
Comment (susiecoleman) [inline conf/CODE.conf]: fixed

File: app/com/gu/identity/frontend/controllers/RegisterAction.scala [hunk 4]
```diff
@@ -63,4 +76,64 @@ class RegisterAction(identityService: IdentityService, val messagesApi: Messages
...
+
+  private def checkUserDataIsUnique(error: ServiceError): String = {
+    error.message match {
+      case "Username in use" => "register-error-username-in-use"
+      case "Email in use" => "register-error-email-in-use"
+      case _ => s"register-${error.id}"
+    }
+  }
+
+  private def redirectRegisterSuccess(cookies: Seq[PlayCookie], returnUrl: ReturnUrl, skipConfirmation: Option[Boolean], group: Option[String]) = {
```
Comment (jamespamplin) [inline app/com/gu/identity/frontend/controllers/RegisterAction.scala]: This function builds the redirect urls and the result, and could be separated into two functions. eg:

```scala
def registerSuccessResult( ... ) =
  SeeOther(registerSuccessRedirectUrl)
    .withCookies(cookies: _*)

def registerSuccessRedirectUrl( ... ) =
  ??? // logic to determine redirect url
```

File: app/com/gu/identity/frontend/controllers/RegisterAction.scala [hunk 5]
```diff
@@ -63,4 +76,64 @@ class RegisterAction(identityService: IdentityService, val messagesApi: Messages
...
+    val groupCode = validateGroupCode(group)
```
Comment (jamespamplin) [inline app/com/gu/identity/frontend/controllers/RegisterAction.scala]: Is it possible to move `validateGroupCode` into the `registerForm` mapping?

File: app/com/gu/identity/frontend/views/models/ErrorViewModel.scala [hunk 6]
```diff
@@ -11,9 +11,13 @@ object ErrorViewModel {
   }
 
   val errorMessages = Map(
-    "error-gateway" -> "There was a problem signing in; please try again.",
-    "error-bad-request" -> "Incorrect email or password; please try again.",
-    "error-registration" -> "Sorry an account could not be set up."
+    "signin-error-gateway" -> "There was a problem signing in; please try again.",
```
Comment (jamespamplin) [inline app/com/gu/identity/frontend/views/models/ErrorViewModel.scala]: Not important, but we should move these to the `messages.en` file. Can be addressed separately as tech debt.

File: app/com/gu/identity/frontend/utils/UrlDecoder.scala [hunk 7]
```diff
@@ -0,0 +1,32 @@
+package com.gu.identity.frontend.utils
+
+object UrlDecoder {
+
+  def getQueryParams(url: String): Map[String, String] = {
```
Comment (jamespamplin) [inline app/com/gu/identity/frontend/utils/UrlDecoder.scala]: Not sure if we should be decoding URL params ourselves. Is there a library we can use instead? Or just use java's built in one? /cc @markjamesbutler @danmassie

File: app/com/gu/identity/frontend/controllers/RegisterAction.scala (hunk 3 above)
Comment (susiecoleman) [inline app/com/gu/identity/frontend/controllers/RegisterAction.scala]: Added

File: app/com/gu/identity/frontend/controllers/RegisterAction.scala (hunk 4 above)
Comment (susiecoleman) [inline app/com/gu/identity/frontend/controllers/RegisterAction.scala]: changed

File: app/com/gu/identity/frontend/controllers/RegisterAction.scala (hunk 5 above)
Comment (susiecoleman) [inline app/com/gu/identity/frontend/controllers/RegisterAction.scala]: Yes, moved

File: app/com/gu/identity/frontend/views/models/ErrorViewModel.scala (hunk 6 above)
Comment (jamespamplin) [inline app/com/gu/identity/frontend/views/models/ErrorViewModel.scala]: Agreed that we can address this as tech debt separately.

File: app/com/gu/identity/frontend/utils/UrlDecoder.scala (hunk 7 above)
Comment (jamespamplin) [inline app/com/gu/identity/frontend/utils/UrlDecoder.scala]: Agreed that we can address this as tech debt.

**Comment by jamespamplin** [comment]
:+1: great work!

File: app/com/gu/identity/frontend/utils/UrlDecoder.scala (hunk 7 above)
Comment (susiecoleman) [inline app/com/gu/identity/frontend/utils/UrlDecoder.scala]: Moved into test directory as currently only used for testing.

**Pull request merged**